pingcap/tiflow
# pkg/ha: refine ha package error information and retry mechnism for etcd operations

### What problem does this PR solve?


Issue Number: close https://github.com/pingcap/tiflow/issues/5235

### What is changed and how it works?

1. Refine retry for dm shard ddl done operation and other etcd operations when dm-master leader is down.
2. Add error info for etcd related errors.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
Test running DM 097 case after this PR, when EOF error occurs dm-worker can automatically resume from error through retry and the error hint is correct.
```bash
$ cat /tmp/dm_test/shardddl3/worker*/log/stdout.log | grep EOF
{"level":"warn","ts":"2022-04-26T15:55:11.170+0800","logger":"etcd-client","caller":"v3@v3.5.2/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc00043bc00/127.0.0.1:8261","attempt":0,"error":"rpc error: code = Unavailable desc = error reading from server: EOF"}
$ cat /tmp/dm_test/shardddl3/worker*/log/dm-worker.log | grep -a EOF
[2022/04/26 17:15:56.360 +08:00] [WARN] [strategy.go:83] ["retry stratey takes effect"] [error="[code=43001:class=unknown error class: 21:scope=internal:level=high], Message: fail to do etcd txn operation: txn commit failed, RawCause: rpc error: code = Unavailable desc = error reading from server: EOF, Workaround: Please check dm-master's node status and the network between this node and dm-master"] [errorVerbose="[code=43001:class=unknown error class: 21:scope=internal:level=high], Message: fail to do etcd txn operation: txn commit failed, RawCause: rpc error: code = Unavailable desc = error reading from server: EOF, Workaround: Please check dm-master's node status and the network between this node and dm-master\ngithub.com/pingcap/tiflow/dm/pkg/terror.(*Error).Delegate\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/pkg/terror/terror.go:312\ngithub.com/pingcap/tiflow/dm/pkg/etcdutil.DoOpsInOneTxnRepeatableWithRetry.func1\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/pkg/etcdutil/etcdutil.go:113\ngithub.com/pingcap/tiflow/dm/pkg/retry.(*FiniteRetryStrategy).Apply\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/pkg/retry/strategy.go:73\ngithub.com/pingcap/tiflow/dm/pkg/etcdutil.DoOpsInOneTxnRepeatableWithRetry\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/pkg/etcdutil/etcdutil.go:110\ngithub.com/pingcap/tiflow/dm/pkg/shardddl/optimism.PutInfo\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/pkg/shardddl/optimism/info.go:183\ngithub.com/pingcap/tiflow/dm/syncer/shardddl.(*Optimist).PutInfo\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/syncer/shardddl/optimist.go:112\ngithub.com/pingcap/tiflow/dm/syncer.(*Syncer).handleQueryEventOptimistic\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/syncer/optimist.go:162\ngithub.com/pingcap/tiflow/dm/syncer.(*Syncer).handleQueryEvent\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/syncer/syncer.go:2772\ngithub.com/pingcap/tiflow/dm/syncer.(*Syncer).Run\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/syncer/syncer.go:2107\ngithub.com/pingcap/tiflow/dm/syncer.(*Syncer).Process\n\t/Users/chauncy/code/goPath/src/github.com/pingcap/tiflow/dm/syncer/syncer.go:696\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1259"] [retry_times=0] [retry_count=5]
```

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression

Related changes

- None

### Release note 

```release-note
improve dm's error hint and robustness for etcd operations
```
